### PR TITLE
review daml-intro-10

### PR DESCRIPTION
**Pull request by garyverhaegen-da**
I started with wanting to remove the warning on incomplete pattern match, thinking example code should be "good". Then I decided I should skim the corresponding docs to check if that warning, or the pragma, were actually mentioned in the text.

It turns out the warning should be there, but the pragma is not explained and not necessary to get the warning. So the pargma can go.

But in the process I noticed a couple things that I belive can be improved in the wording of the document, so I went agead and made a few changes.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: docs/source/daml/intro/10_Functional101.rst [hunk 1]
```diff
@@ -116,29 +116,36 @@ If we want to partially apply an infix operation we can also do that as follows:
...
+Let's turn this into prose: Given that ``t`` is the type of a template, and that ``t`` has a choice ``c`` with return type ``r``, the ``exercise`` function maps a ``ContractId`` for a contract of type ``t`` to a function that takes the choice arguments of type ``c`` and returns an ``Update`` resulting in type ``r``.
 
 That's quite a mouthful, and does require one to know what *meaning* the typeclass ``Choice`` gives to parameters ``t`` ``c`` and ``r``, but in many cases, that's obvious from the context or names of typeclasses and variables.
 
+Note that using single letters, while common, is not mandatory. The above may be made a little bit clearer by expanding the type parameter names, at the cost of making the code a bit longer:
+
+.. code-block:: daml
+
+  exercise : (Template template, Choice template choice result) =>
+               ContractId template -> choice -> Update result
```
Comment (stefanobaghino-da) [inline docs/source/daml/intro/10_Functional101.rst:143]: I like this a lot. Single letters and unreadable shortenings are terrible.

LGTM.

:trollface:

**Review by stefanobaghino-da — COMMENTED** [review]
LGTM but I'll leave it for @carrielaben-da to judge the changes.

File: docs/source/daml/intro/daml/daml-intro-10/daml/Main.daml [hunk 2]
```diff
@@ -35,7 +34,7 @@ increment2 = add 1
 
 -- INFIX2_BEGIN
 increment3 = (1 +)
-decrement = (- 1)
```
Comment (garyverhaegen-da) [inline docs/source/daml/intro/daml/daml-intro-10/daml/Main.daml:38]: `(- 1)`, in Daml as well as in Haskell, is a `Int`, not a `Int -> Int`.

File: docs/source/daml/intro/10_Functional101.rst [hunk 3]
```diff
@@ -190,29 +203,29 @@ You have probably already guessed it: Anywhere you can put a value in Daml you c
   :start-after: -- FUNCTION_IN_DATA_BEGIN
   :end-before: -- FUNCTION_IN_DATA_END
 
-More commonly, it makes sense to define functions locally, inside a ``let`` clause or similar. A good example of this are the ``validate`` and ``transfer`` functions defined locally in the ``Trade_Settle`` choice of the  model from chapter 8:
+More commonly, it makes sense to define functions locally, inside a ``let`` clause or similar. A good example of this are the ``validate`` and ``transfer`` functions defined locally in the ``Trade_Settle`` choice of the  model from chapter 9:
```
Comment (carrielaben-da) [inline docs/source/daml/intro/10_Functional101.rst:206]:
```suggestion
More often it makes sense to define functions locally, inside a ``let`` clause or similar. Good examples of this are the ``validate`` and ``transfer`` functions defined locally in the ``Trade_Settle`` choice of the  model from chapter 9:
```

File: docs/source/daml/intro/10_Functional101.rst [hunk 4]
```diff
@@ -190,29 +203,29 @@ You have probably already guessed it: Anywhere you can put a value in Daml you c
...
 
 .. note::
 
-    Bear in mind that functions are not serializable, so you can't use them inside template arguments, or as choice in- or outputs. They also don't have instances of the ``Eq`` or ``Show`` typeclasses which one would commonly want on data types.
+    Bear in mind that functions are not serializable, so you can't use them inside template arguments, as choice inputs, or as choice outputs. They also don't have instances of the ``Eq`` or ``Show`` typeclasses which one would commonly want on data types.
 
-You can probably guess what the ``mapA`` and ``mapA_``\ s in the above choice do. They somehow loop through the lists of assets, and approvals, and the functions ``validate`` and ``transfer`` to each, performing the resulting ``Update`` action in the process. We'll look at that more closely under :ref:`looping` below.
+The ``mapA`` and ``mapA_`` functions loop through the lists of assets and approvals, and apply the functions ``validate`` and ``transfer`` to each element of those lists, performing the resulting ``Update`` action in the process. We'll look at that more closely under :ref:`looping` below.
```
Comment (carrielaben-da) [inline docs/source/daml/intro/10_Functional101.rst:223]:
```suggestion
The ``mapA`` and ``mapA_`` functions loop through the lists of assets and approvals and apply the functions ``validate`` and ``transfer`` to each element of those lists, performing the resulting ``Update`` action in the process. We'll look at that more closely under :ref:`looping` below.
```

File: docs/source/daml/intro/10_Functional101.rst [hunk 5]
```diff
@@ -427,26 +440,26 @@ Daml also has an operator ``(!!) : [a] -> Int -> a`` which returns an element in
...
+In effect, the lambda passed to ``foldl`` only "wants" to act on a single element of the (zipped-up) input list, but still has to manage the concatenation of the whole state. Acting on each element separately is a common-enough pattern that there is a specialized function for it: ``map : (a -> b) -> [a] -> [b]``. Using it, we can rewrite ``sumArr`` to:
 
 .. literalinclude:: daml/daml-intro-10/daml/Main.daml
   :language: daml
   :start-after: -- SUM_ARR3_BEGIN
   :end-before: -- SUM_ARR3_END
 
-As a rule of thumb: Use ``map`` if the result has the same shape as the input and you don't need to carry state from one iteration to the next. Use folds if you need to accumulate state in any way.
+As a rule of thumb, use ``map`` if the result has the same shape as the input and you don't need to carry state from one iteration to the next. Use folds if you need to accumulate state in any way.
```
Comment (carrielaben-da) [inline docs/source/daml/intro/10_Functional101.rst:462]:
```suggestion
As a rule, use ``map`` if the result has the same shape as the input and you don't need to carry state from one iteration to the next. Use folds if you need to accumulate state in any way.
```

File: docs/source/daml/intro/10_Functional101.rst [hunk 6]
```diff
@@ -463,25 +476,25 @@ You may be tempted to make ``reverseWorker`` a local definition inside ``reverse
 Folds and Maps in Action Contexts
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-The folds and ``map`` function above are pure in the sense introduced in :doc:`5_Restrictions`: The functions used to map or process items have no side-effects. In day-to-day Daml that's the exception rather than the rule. If you have looked at the chapter 8 models, you'll have noticed ``mapA``, ``mapA_``, and ``forA`` all over the place. A good example are the ``mapA`` in the ``testMultiTrade`` script:
+The folds and ``map`` function above are pure in the sense introduced in :doc:`5_Restrictions`: The functions used to map or process items have no side-effects. If you have looked at the chapter 9 models, you'll have noticed ``mapA``, ``mapA_``, and ``forA``, that seem to serve a similar role but within ``Action``\ s . A good example is the ``mapA`` call in the ``testMultiTrade`` script:
```
Comment (carrielaben-da) [inline docs/source/daml/intro/10_Functional101.rst:479]:
```suggestion
The folds and ``map`` function above are pure in the sense introduced in :doc:`5_Restrictions`: The functions used to map or process items have no side effects. If you have looked at the chapter 9 models, you'll have noticed ``mapA``, ``mapA_``, and ``forA``, which seem to serve a similar role but within ``Action``\ s . A good example is the ``mapA`` call in the ``testMultiTrade`` script:
```

File: docs/source/daml/intro/10_Functional101.rst [hunk 7]
```diff
@@ -463,25 +476,25 @@ You may be tempted to make ``reverseWorker`` a local definition inside ``reverse
...
+The ``A`` in ``mapA`` stands for "Action", and you'll find that many functions that have something to do with "looping" have an ``A`` equivalent. The most fundamental of all of these is ``foldlA : Action m => (b -> a -> m b) -> b -> [a] -> m b``, a left fold with side effects. Here the inner function has a side-effect indicated by the ``m`` so the end result ``m b`` also has a side effect: the sum of all the side effects of the inner function.
 
-Have a go at implementing ``foldlA`` in terms of ``foldl`` and ``sequence`` and ``mapA`` in terms of ``foldA``. Here are some possible implementations:
+To improve your familiarity with these concepts, you could try implementing ``foldlA`` in terms of ``foldl``, as well as ``sequence`` and ``mapA`` in terms of ``foldlA``. Here is one set of possible implementations:
```
Comment (carrielaben-da) [inline docs/source/daml/intro/10_Functional101.rst:497]:
```suggestion
To improve your familiarity with these concepts, try implementing ``foldlA`` in terms of ``foldl``, as well as ``sequence`` and ``mapA`` in terms of ``foldlA``. Here is one set of possible implementations:
```

**Review by carrielaben-da — APPROVED** [review]
Good stuff - I took the opportunity to add a few more tweaks.

**Pull request merged**